### PR TITLE
Nifi 4908 rebase

### PR DESCRIPTION
**Pull request by david-streamlio**
Thank you for submitting a contribution to Apache NiFi.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with NIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [x] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi folder?
- [x] Have you written or updated unit tests to verify your changes?
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the LICENSE file, including the main LICENSE file under nifi-assembly?
- [ ] If applicable, have you updated the NOTICE file, including the main NOTICE file found under nifi-assembly?
- [ ] If adding new Properties, have you added .displayName in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-processors/src/main/java/org/apache/nifi/processors/pulsar/PublishPulsar_1_0.java [hunk 1]
```diff
@@ -0,0 +1,373 @@
...
+
+        if (StringUtils.isBlank(topic)) {
+            logger.error("Invalid topic specified {}", new Object[] {topic});
+            session.transfer(flowFile, REL_FAILURE);
+            return;
+        }
+
+        // Read the contents of the FlowFile into a byte array
+        final byte[] messageContent = new byte[(int) flowFile.getSize()];
+        session.read(flowFile, new InputStreamCallback() {
```
[the remark this answers is not in the document]
Comment (MikeThomsen) [reply nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-processors/src/main/java/org/apache/nifi/processors/pulsar/PublishPulsar_1_0.java]: Alternatively, you can use `session.exportTo` to stream the contents into `ByteArrayOutputStream`. Your call.

**Comment by MikeThomsen** [comment]
You should strongly consider setting up some integration tests that can be run against a simple Docker image. All you have to do with NiFi to get that done is add a few classes with "IT" at the end of their name and you can run them with `mvn integration-test -Pintegration-tests`.

**Comment by david-streamlio** [comment]
@MikeThomsen   Is there another car bundle that uses a Docker image for integration testing that I can use as an example?   

I will clean up the formatting / style errors.

**Comment by david-streamlio** [comment]
@MikeThomsen   We have a pre-configured Docker image that has Pulsar if you want to use that for testing.  See https://streaml.io/docs/getting-started. for detailed instructions.

**Comment by MikeThomsen** [comment]
@david-streamlio thanks for the link.

> Is there another car bundle that uses a Docker image for integration testing that I can use as an example?

The Mongo package, sorta. We wrote all of the tests to assume Mongo defaults, so if you set up a simple Docker install of Mongo, all of the defaults will just click between the image and the int tests.

File: nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-client-service/src/main/java/org/apache/nifi/pulsar/StandardPulsarClientPool.java [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+        builder.append(uri);
+        return builder.toString();
+    }
+
+    private ClientConfiguration getClientConfig(ConfigurationContext context) throws UnsupportedAuthenticationException {
+
+        if (clientConfig == null) {
+            clientConfig = new ClientConfiguration();
+
+            if (context.getProperty(CONCURRENT_LOOKUP_REQUESTS).isSet()) {
```
[the remark this answers is not in the document]
Comment (david-streamlio) [reply nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-client-service/src/main/java/org/apache/nifi/pulsar/StandardPulsarClientPool.java]: It is my intent to use the default values for org.apache.pulsar.client.api.ClientConfiguration, if and only if the user didn't explicitly override them via one of these properties.  The default values I provide for each of these is directly from the API default values....

**Comment by MikeThomsen** [comment]
@david-streamlio I'm going to try to get back to this very soon.

**Comment by david-streamlio** [comment]
Thanks for the update.....I have merged my changes that came from your code review into the NIFI-4914-rebase PR as well.  That PR has extends the base Pulsar processor capabilities to include RecordBased processing as well

**Comment by MikeThomsen** [comment]
Ok. If you merged 4908 into 4914, go ahead and close out this out. Also, it would be a good idea to rebase 4914 and squash it into one commit starting out for a code review.

**Comment by MikeThomsen** [comment]
Read [this blog post](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html). It goes into good detail about squashing commits during a rebase, along with the caveats in doing so. Basically what you do is follow those directions (pretty simple, but worth it for the background) and then do `git push origin BRANCH_HOME --force` to override the existing remote version of 4914 once you have rebased and squashed.

**Comment by MikeThomsen** [comment]
The rule of thumb there is to be patient and do it multiple steps with as many commits as you have in these two branches. I'd recommend only squashing a few commits together a time until 4914 is one pristine commit ready for review.

**Comment by david-streamlio** [comment]
So the goal is to get my 23 commits squashed down into a single commit?

**Comment by MikeThomsen** [comment]
Yeah. Since this is your first time, I'd recommend backing up the repo before trying it. Once you've gotten them down to one commit, it won't need to be done again until the review process is done.

**Comment by david-streamlio** [comment]
Alright, sounds good.

**Pull request closed without merging**